### PR TITLE
Updating profiler similar to upstream for CUDA

### PR DESCRIPTION
**Pull request by reza-amd**
This PR solves the following problem with the profiler:

- Assign correct device ID to events
- Add host events to the trace view
- Add new APIs to the tracking list
- Solve wrong profile views in the TensorBoard

### CODE REVIEW

**Comment by reza-amd** [comment]
jenkins : retest this please

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h [hunk 1]
```diff
@@ -61,6 +61,7 @@ namespace wrap {
 
 #endif  // PLATFORM_GOOGLE
 
+//BUG(rocm-profiler): roctracer_enable_op_activity does not exist in rocm > 4.0.1
```
Comment (deven-amd) [inline tensorflow/stream_executor/rocm/roctracer_wrapper.h]: if we no longer call `roctracer_enable_op_activity`, simply remove all references to it

File: tensorflow/core/profiler/protobuf/hardware_types.proto [hunk 2]
```diff
@@ -26,5 +26,6 @@ message DeviceCapabilities {
   uint32 num_cores = 2;
   uint64 memory_size_in_bytes = 3;
   uint64 memory_bandwidth = 4;  // Bytes/s.
-  CudaComputeCapability compute_capability = 5;
+  GPUComputeCapability compute_capability = 5;
```
Comment (deven-amd) [inline tensorflow/core/profiler/protobuf/hardware_types.proto:29]: think we cannot use the `major,minor` int tuple to represent AMDGPU capabilities....we are going to need to use a string instead.

see here - https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/gfx90a_support/tensorflow/compiler/xla/service/gpu/gpu_types.h#L27-L38 

we need to do something similar here -- create a new message `GPUComputeCapability` which is a union of `CudaComputeCapability` amd `ROCmComputeCapability` (which will just be a string for now)

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 3]
```diff
@@ -15,6 +16,11 @@ limitations under the License.
 
 #include "tensorflow/core/profiler/internal/gpu/rocm_tracer.h"
 
+#include <chrono>
+#include <iostream>
+#include <sstream>
+#include <thread>
+
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: these header includes should not be here. they were removed because the upstream reviewers wanted them gone.  there are other changes (using `uint32_t` instead `uint32`) that have also crept back in.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 4]
```diff
@@ -258,11 +270,10 @@ const char* GetRocmTracerEventDomainName(const RocmTracerEventDomain& domain) {
...
 }
 
-void DumpRocmTracerEvent(const RocmTracerEvent& event,
-                         uint64_t start_walltime_ns,
-                         uint64_t start_gputime_ns) {
+void DumpRocmTracerEvent(const RocmTracerEvent& event, uint64 start_walltime_ns,
+                         uint64 start_gputime_ns, const string& message) {
   std::ostringstream oss;
-
+
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: nit : drop the extra line

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 5]
```diff
@@ -800,58 +1209,128 @@ void RocmActivityCallbackImpl::AddHipMallocEvent(
...
+   activity record contains device/stream ID
+ */
   RocmTracerEvent event;
   event.domain = RocmTracerEventDomain::HCC_OPS;
   event.type = RocmTracerEventType::Kernel;
   event.source = RocmTracerEventSource::Activity;
-  event.name =
+  // event.name =
+  //    wrap::roctracer_op_string(record->domain, record->op, record->kind);
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: what is the reason for commenting this out? if not needed, remove it completely

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 6]
```diff
@@ -800,58 +1209,128 @@ void RocmActivityCallbackImpl::AddHipMallocEvent(
...
+  event.type = RocmTracerEventType::Synchronization;
   event.source = RocmTracerEventSource::Activity;
   event.name =
       wrap::roctracer_op_string(record->domain, record->op, record->kind);
   event.correlation_id = record->correlation_id;
   event.annotation = collector_->annotation_map()->LookUp(event.correlation_id);
-
   event.start_time_ns = record->begin_ns;
-  event.end_time_ns = record->end_ns;
+  event.end_time_ns = std::max(record->end_ns, record->begin_ns + 1);
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:1220]: add a comment here as to why this is needed

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 7]
```diff
@@ -800,58 +1209,128 @@ void RocmActivityCallbackImpl::AddHipMallocEvent(
       wrap::roctracer_op_string(record->domain, record->op, record->kind);
   event.correlation_id = record->correlation_id;
   event.annotation = collector_->annotation_map()->LookUp(event.correlation_id);
-
+  // similar to CUDA we set this to the default stream
+  event.stream_id = 0;
   event.start_time_ns = record->begin_ns;
-  event.end_time_ns = record->end_ns;
+  event.end_time_ns = std::max(record->end_ns, record->begin_ns + 1);
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:1194]: add a comment here as to why this is needed

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 8]
```diff
@@ -677,84 +1029,125 @@ Status RocmActivityCallbackImpl::operator()(const char* begin,
...
-      event.type = RocmTracerEventType::MemcpyD2H;
-      event.memcpy_info.async = false;
-      break;
     case HIP_API_ID_hipMemcpyDtoHAsync:
       event.type = RocmTracerEventType::MemcpyD2H;
-      event.memcpy_info.async = true;
+      event.memcpy_info.async =
+          (record->op == HIP_API_ID_hipMemcpyDtoHAsync) ? true : false;
+      // event.name =
+      // GetRocmTracerEventTypeName(RocmTracerEventType::MemcpyD2H);
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: delete lines

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 9]
```diff
@@ -677,84 +1029,125 @@ Status RocmActivityCallbackImpl::operator()(const char* begin,
...
-      event.type = RocmTracerEventType::MemcpyH2D;
-      event.memcpy_info.async = false;
-      break;
     case HIP_API_ID_hipMemcpyHtoDAsync:
       event.type = RocmTracerEventType::MemcpyH2D;
-      event.memcpy_info.async = true;
+      event.memcpy_info.async =
+          (record->op == HIP_API_ID_hipMemcpyHtoDAsync) ? true : false;
+      // event.name =
+      // GetRocmTracerEventTypeName(RocmTracerEventType::MemcpyH2D);
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: delete lines....same applies in all other case statements

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 10]
```diff
@@ -677,84 +1029,125 @@ Status RocmActivityCallbackImpl::operator()(const char* begin,
...
+      // GetRocmTracerEventTypeName(RocmTracerEventType::MemcpyOther);
+      break;
+    case HIP_API_ID_hipMemcpyPeerAsync:
+      event.type = RocmTracerEventType::MemcpyP2P;
       event.memcpy_info.async = true;
-      // ROCM TODO: figure out a way to properly populate this field.
-      event.memcpy_info.destination = record->device_id;
+      // event.name =
+      // GetRocmTracerEventTypeName(RocmTracerEventType::MemcpyD2H);
       break;
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: keep same pattern here as for all the other case statements above

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 11]
```diff
@@ -609,63 +916,108 @@ Status RocmActivityCallbackImpl::operator()(const char* begin,
...
             AddHccKernelActivityEvent(record);
             tracer_->RemoveFromPendingActivityRecords(record->correlation_id);
             break;
           case HIP_OP_ID_COPY:
-            DumpActivityRecord(record);
-            AddHccMemcpyActivityEvent(record);
+            switch (record->kind) {
+              case 4595: /*CopyDeviceToHost*/
+              case 4596:
+              case 4597: { /*CopyHostToDevice*/
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: can we use enum literals/#defines associated with the numbers instead....magic numbers are always a bad idea

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 12]
```diff
@@ -446,153 +550,356 @@ void RocmApiCallbackImpl::AddKernelEventUponApiExit(
...
+  event.start_time_ns = enter_time;
+  event.end_time_ns = exit_time;
 
-  collector_->AddEvent(std::move(event));
+  switch (cbid) {
+    case HIP_API_ID_hipStreamSynchronize: {
+      event.synchronization_info.sync_type =
+          RocmTracerSyncTypes::StreamSynchronize;
+      const hipStream_t& stream = data->args.hipStreamSynchronize.stream;
+      event.device_id = hipGetStreamDeviceId(stream);
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:856]: `wrap::hipGetStreamDeviceId` ? same applies to all other HIP API calls

Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:856]: also does calling HIP APIs here trigger a callback if that API is being tracked? If yes, how do we handle that situation?

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 13]
```diff
@@ -446,153 +550,356 @@ void RocmApiCallbackImpl::AddKernelEventUponApiExit(
...
+  destenation as the source device id). Ultimately, to figure out the actual
+  device we can use hipPointerGetAttributes but we do not do that now .In the
+  other words, we assume we use hipMemcpyPeer to achieve the copy between
+  devices.
+  */
 
-  // TODO(rocm): figure out a way to properly populate this field.
-  event.memcpy_info.destination = 0;
   switch (cbid) {
     case HIP_API_ID_hipMemcpyDtoH:
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: no bracket for this case, but there is one for all others...please run clang-format on the changes

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 14]
```diff
@@ -420,6 +515,7 @@ void RocmApiCallbackImpl::AddKernelEventUponApiExit(
       event.kernel_info.block_x = data->args.hipHccModuleLaunchKernel.blockDimX;
       event.kernel_info.block_y = data->args.hipHccModuleLaunchKernel.blockDimY;
       event.kernel_info.block_z = data->args.hipHccModuleLaunchKernel.blockDimZ;
+#if TF_ROCM_VERSION >= 30800
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: this is always true....we can get rid of that #else block. this looks like another case, of merging gone wrong resulting in removed changes creeping back in

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h [hunk 15]
```diff
@@ -221,17 +243,30 @@ class RocmApiCallbackImpl {
...
+                                            uint64_t exit_time);
+  void AddSynchronizeEventUponApiExit(uint32_t cbid, const hip_api_data_t* data,
+                                      uint64_t enter_time, uint64_t exit_time);
 
   RocmTracerOptions options_;
   RocmTracer* tracer_ = nullptr;
   RocmTraceCollector* collector_ = nullptr;
+  mutex api_call_start_mutex_;
+  // TODO(rocm-profiler): replace this with absl hashmap
+  std::map<uint32_t, uint64_t> api_call_start_time_ TF_GUARDED_BY(api_call_start_mutex_);  // keep a map from the corr. id to start time
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.h]: add a comment here describing what the map is for (`corelation_id --> start_time` ? ) and why it is needed (check whether we received the "enter api" callback ? )

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h [hunk 16]
```diff
@@ -221,17 +243,30 @@ class RocmApiCallbackImpl {
   Status operator()(uint32_t domain, uint32_t cbid, const void* cbdata);
 
  private:
-  void AddKernelEventUponApiExit(uint32_t cbid, const hip_api_data_t* data);
-  void AddMemcpyEventUponApiExit(uint32_t cbid, const hip_api_data_t* data);
-  void AddMemsetEventUponApiExit(uint32_t cbid, const hip_api_data_t* data);
-  void AddMallocEventUponApiExit(uint32_t cbid, const hip_api_data_t* data);
+  void AddKernelEventUponApiExit(uint32_t cbid, const hip_api_data_t* data,
+                                 uint64_t enter_time, uint64_t exit_time);
+  void AddNormalMemcpyEventUponApiExit(uint32_t cbid, const hip_api_data_t* data,
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.h]: why `Normal` in the name?

**Review by deven-amd — COMMENTED** [review]
have not looked at the changes in `device_trace_rocm.cc` in detail, but am going to assume they are okay.

LGTM other the minor nits I have put comments for.

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h (hunk 1 above)
Comment (reza-amd) [reply tensorflow/stream_executor/rocm/roctracer_wrapper.h]: will be done on the next commit.

File: tensorflow/core/profiler/protobuf/hardware_types.proto (hunk 2 above)
Comment (reza-amd) [reply tensorflow/core/profiler/protobuf/hardware_types.proto:29]: In our data structure returned from "hipGetDeviceProperties" API call (https://rocm-developer-tools.github.io/HIP/structhipDeviceProp__t.html), we have major and minor. We also have a gcnArch field (an integer keeping full gfx number). This formula is the relation between them (major * 100) + (minor % 10) = gcnArch - (gcnArch % 10). In terms of maintainability, I believe the "hipDeviceProp_t" will not change frequently. In addition, I am afraid the gpu_types may abstract the real hardware capability in future and for profiling reason I think it is better to stick with the what hip API returns about the GPU. Considering this, if you think it is still better to use the device capability representation in XLA, I will change it.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 3 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: Done.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 4 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: Done.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 5 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: Done.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 6 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:1220]: Done.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 7 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:1194]: Done.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 8 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: Done.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 9 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: Done.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 10 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: Done.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 13 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: Done.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 11 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: I agree, but I could not found their associated enum in. I will add a TODO for this.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 12 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:856]: I have tried to add hipGetStreamDeviceId in the "wrap" namespace of rocm driver API header file. I have not been successful in compiling it in this case. I also saw "hipKernelNameRef" being used directly a few lines before without wrapping, so I decided to call it directly. I will make a TODO for it. 

This is a good point, in that case, we will have an infinite circle of calls. Currently, "hipGetStreamDeviceId" is not in the list of APIs we track for HIP API callbacks.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h (hunk 15 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h]: This is needed to track the HIP API calls from the host side. In the enter callback, we sample the current time, and fill this map  based on the correlation ID. At exit time, we use this map to retrieve the enter time and make a host event with these timing.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h (hunk 16 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h]: This is the convention used for CUDA to differentiate them from the peer memory copies.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 14 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: Done!

File: tensorflow/core/profiler/protobuf/hardware_types.proto (hunk 2 above)
Comment (deven-amd) [reply tensorflow/core/profiler/protobuf/hardware_types.proto:29]: think we should use this field - https://github.com/ROCm-Developer-Tools/HIP/blob/main/include/hip/hip_runtime_api.h#L118

that is what I had in mind when I proposed to use `string` for AMDGPU name. This string contains "gfxNNN" as well as any other device subfeatures (e.g. `gfx908:sramecc+:xnack-` )

you can make this change as a separate followup PR to this one

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 17]
```diff
@@ -1108,7 +1574,7 @@ Status RocmTracer::DisableActivityTracing() {
             << ", Threshold = " << threshold;
     VLOG(3) << "Wait for pending activity records : sleep for " << duration_ms
             << " ms";
-    tensorflow::profiler::SleepForMillis(duration_ms);
+    std::this_thread::sleep_for(std::chrono::milliseconds(duration_ms));
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: this is a merge error, undo this change

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc [hunk 18]
```diff
@@ -26,7 +26,6 @@ limitations under the License.
 #include "tensorflow/core/platform/macros.h"
 #include "tensorflow/core/platform/mem.h"
 #include "tensorflow/core/profiler/internal/cpu/annotation_stack.h"
-#include "tensorflow/core/profiler/utils/time_utils.h"
```
Comment (deven-amd) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: this is a merge error, undo this change

**Review by deven-amd — APPROVED** [review]
just a couple of minor changes needed....you can do them in this PR itself or in a followup PR.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 18 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: Done.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc (hunk 17 above)
Comment (reza-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: Done.
